### PR TITLE
update haskell backends selection flow to match changes in 74fdbb67

### LINKED ISSUE #7426: Company backends not loading in haskell layer
#### Description :octocat:

Company backends are not loaded in haskell layer.

For the reproduction steps, I used the `ghci` backend since its got less pre-reqs,

But I actually use the `ghc-mod` backend (i.e i actually have `(haskell :variables haskell-completion-backend 'ghc-mod)`)

And I can see that [this](https://github.com/syl20bnr/spacemacs/blob/master/layers/+lang/haskell/funcs.el#L26-L43) function is being run
because the major mode key bindings are added to the `SPC` menu.

I don't know to reproduce it, but i have noticed under some rare conditions, the additional backends (`company-{ghc|ghci|yasnippet}`) do become active,
But they never seem to be active on startup, if they ever do get triggered, its always sometime in the middle of a session 
#### Reproduction guide :beetle:
1. Activate `haskell` layer in `.spacemacs` with line `(haskell :variables haskell-completion-backend 'ghci)`
2. Start Emacs
3. Load haskell source file inside haskell project
4. move cursor to the end of any word
5. run `M-x company-diag`
6. observe that neither `company-ghci` nor `company-yasnippet` are in the list.
7. open up another haskell source file in the project 
8. repeat 4 and 5
9. notice that `company-ghci` and `company-yasnippet` now appear in the backends list

_Observed behaviour:_ :eyes: :broken_heart:

Neither `company-ghci` nor `company-yasnippet` appear in the list of backends active

_Expected behaviour:_ :heart: :smile:

Expect to see `company-ghci` and `company-yasnippet` in the list of backends active/queried.
#### System Info :computer:
- OS: darwin
- Emacs: 24.5.2
- Spacemacs: 0.200.0
- Spacemacs branch: master (rev. e6357a5)
- Graphic display: nil
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:

``` elisp
osx
auto-completion
chrome
deft
themes-megapack
git
version-control
restclient
syntax-checking
search-engine
c-c  
markdown
org
(scala :variables scala-enable-eldoc nil)
(haskell :variables haskell-completion-backend 'ghci)
latex
shell-scripts
shell
sql
```

### PR DESCRIPTION
**Pull request by d12frosted**
It's possible that I have over-engineered it. Just let me know if something is wrong. What I don't like is how `haskell-packages` became bloated! 😿 

Additional notes:

- Do not install packages that are not used (like `intero` users don't have to install `dante` or `ghc`).
- Defer `dante`.

---

Also fixes #7426

### CODE REVIEW

File: layers/+lang/haskell/packages.el [hunk 1]
```diff
@@ -12,27 +12,61 @@
...
     hlint-refactor
-    intero
-    (dante :toggle (version<= "25" emacs-version))
     ))
 
 (defun haskell/init-cmm-mode ()
   (use-package cmm-mode
     :defer t))
 
+(defun haskell/post-init-company ()
```
Comment (beta1440) [inline layers/+lang/haskell/packages.el]: Why do you call `spacemacs|add-company-backends` here instead of the `haskell-init-company-*backend*` functions?

**Comment by d12frosted** [comment]
@beta1440 thanks for noticing, I've fixed that.

**Comment by wedens** [comment]
Any reason this isn't merged yet?

File: layers/+lang/haskell/packages.el [hunk 2]
```diff
@@ -12,59 +12,163 @@
...
     hlint-refactor
-    intero
-    (dante :toggle (version<= "25" emacs-version))
     ))
 
 (defun haskell/init-cmm-mode ()
   (use-package cmm-mode
     :defer t))
 
+(defun haskell/post-init-company ())
```
[the remark this answers is not in the document]
Comment (d12frosted) [reply layers/+lang/haskell/packages.el]: That's a good question!

**Comment by d12frosted** [comment]
Thank you :+1: Cherry-picked into `develop`, you can safely delete your branch.

Resulting commit is 152d41e7c.

CC @wedens

**Comment by d12frosted** [comment]
P. S. my next goal is to make it locally configurable.

**Pull request closed without merging**